File: impl/ts/tests/correlated_pushdown.test.ts
// Correlated primary-key pushdown over a MULTI-LEAF inner table (spec/design/cost.md §3 "bounded
// scan / correlated"). The conformance corpus
// (spec/conformance/suites/subquery/correlated_pushdown.test) pins the cost contract on single-leaf
// tables; this exercises what it cannot — an inner table wide enough that re-scanning it per outer row
// would be visibly expensive, so the per-outer-row seek is the difference between sublinear and
// quadratic. The win is shown by contrast: `inner.pk = o.col` (bounded) vs `inner.v = o.col` (a full
// re-scan), which return the SAME rows because v == id.

import assert from "node:assert/strict";
import { test } from "node:test";
import { Database, execute } from "../src/lib.ts";
import type { Value } from "../src/value.ts";

// `o` is five outer rows whose k-values all exist as inner ids; `inr` is n rows (id int32 PRIMARY KEY,
// v int32; v == id) wide enough to span several leaves.
function correlatedTables(n: number): Database {
  const db = new Database();
  execute(db, "CREATE TABLE o (id int32 PRIMARY KEY, k int32)");
  execute(db, "CREATE TABLE inr (id int32 PRIMARY KEY, v int32)");
  execute(db, "INSERT INTO o VALUES (1, 100), (2, 300), (3, 500), (4, 700), (5, 900)");
  const parts: string[] = [];
  for (let i = 1; i <= n; i++) parts.push(`(${i},${i})`);
  execute(db, "INSERT INTO inr VALUES " + parts.join(","));
  return db;
}

function cost(db: Database, sql: string): bigint {
  return execute(db, sql).cost;
}

function intOf(v: Value): number {
  if (v.kind !== "int") throw new Error("expected int, got " + v.kind);
  return Number(v.int);
}

function ids(db: Database, sql: string): number[] {
  const o = execute(db, sql);
  if (o.kind !== "query") throw new Error("expected a query result");
  return o.rows.map((r) => intOf(r[0]!));
}

test("correlated EXISTS seek is sublinear", () => {
  const db = correlatedTables(1000);
  // Both correlate the inner to each outer row; `inr.id` is the PK (seeks), `inr.v` is not (full
  // re-scan). v == id, so they select the SAME inner rows and the SAME outer rows survive.
  const bounded = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.id = o.k)";
  const unbounded = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.v = o.k)";
  assert.deepStrictEqual(ids(db, bounded), [1, 2, 3, 4, 5]);
  assert.deepStrictEqual(ids(db, unbounded), [1, 2, 3, 4, 5]);

  const seek = cost(db, bounded);
  const scan = cost(db, unbounded);
  // The non-PK correlation re-scans all ~1000 inner rows for each of the 5 outer rows; the PK
  // pushdown seeks instead, so it is an order of magnitude cheaper.
  assert.ok(seek * 10n < scan, `correlated seek ${seek} should be far below full re-scan ${scan}`);
  // Sublinear in the inner size: 5 outer rows, each ≈ a point lookup (height + a row), not ~1000.
  assert.ok(seek <= 400n, `correlated seek ${seek} should be sublinear (≈ outer × height), not ~5000`);
});

test("correlated scalar seek matches unbounded rows", () => {
  const db = correlatedTables(1000);
  // A correlated SCALAR subquery seeking the inner PK returns each outer row's inner value. Rows are
  // identical to what a full re-scan would produce; only the cost differs.
  const bounded = "SELECT (SELECT inr.v FROM inr WHERE inr.id = o.k) FROM o ORDER BY o.id";
  const unbounded = "SELECT (SELECT inr.v FROM inr WHERE inr.v = o.k) FROM o ORDER BY o.id";
  assert.deepStrictEqual(ids(db, bounded), [100, 300, 500, 700, 900]);

  const seek = cost(db, bounded);
  const scan = cost(db, unbounded);
  assert.ok(seek * 10n < scan, `correlated scalar seek ${seek} should be far below full re-scan ${scan}`);
});

test("correlated miss and NULL outer seek nothing", () => {
  const db = correlatedTables(1000);
  // An outer k with no matching inner id is a point-lookup miss (visits the leaf, reads no row); a
  // NULL outer k is a 3VL-empty bound (reads no page, no row). Neither re-scans the inner.
  execute(db, "INSERT INTO o VALUES (6, 999999), (7, NULL)");
  const q = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.id = o.k)";
  assert.deepStrictEqual(ids(db, q), [1, 2, 3, 4, 5]);
  assert.ok(cost(db, q) <= 500n, "seek cost should stay sublinear with a miss and a NULL outer row");
});

File: impl/go/correlated_pushdown_test.go
package jed

// Correlated primary-key pushdown over a MULTI-LEAF inner table (spec/design/cost.md §3 "bounded
// scan / correlated"). The conformance corpus
// (spec/conformance/suites/subquery/correlated_pushdown.test) pins the cost contract on single-leaf
// tables; this exercises what it cannot — an inner table wide enough that re-scanning it per outer row
// would be visibly expensive, so the per-outer-row seek is the difference between sublinear and
// quadratic. The win is shown by contrast: `inner.pk = o.col` (bounded) vs `inner.v = o.col` (a full
// re-scan), which return the SAME rows because v == id.

import (
	"fmt"
	"strings"
	"testing"
)

// correlatedTables builds `o` (five outer rows whose k-values all exist as inner ids) and `inr` (n
// rows id i32 PRIMARY KEY, v i32; v == id) large enough to span several leaves.
func correlatedTables(t *testing.T, n int) *Session {
	t.Helper()
	var b strings.Builder
	b.WriteString("INSERT INTO inr VALUES ")
	for i := 1; i <= n; i++ {
		if i > 1 {
			b.WriteByte(',')
		}
		fmt.Fprintf(&b, "(%d,%d)", i, i)
	}
	return dbWith(
		t,
		"CREATE TABLE o (id i32 PRIMARY KEY, k i32)",
		"CREATE TABLE inr (id i32 PRIMARY KEY, v i32)",
		"INSERT INTO o VALUES (1, 100), (2, 300), (3, 500), (4, 700), (5, 900)",
		b.String(),
	)
}

func mustIds(t *testing.T, db dbHandle, sql string) []int64 {
	t.Helper()
	out, err := queryOutcome(db, sql, nil)
	if err != nil {
		t.Fatalf("%s: unexpected error %v", sql, err)
	}
	got := make([]int64, len(out.Rows))
	for i, r := range out.Rows {
		got[i] = r[0].Int
	}
	return got
}

func TestCorrelatedExistsSeekIsSublinear(t *testing.T) {
	const n = 1000
	db := correlatedTables(t, n)

	// Both correlate the inner to each outer row; `inr.id` is the PK (seeks), `inr.v` is not (full
	// re-scan). v == id, so they select the SAME inner rows and the SAME outer rows survive.
	const bounded = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.id = o.k)"
	const unbounded = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.v = o.k)"
	if got := mustIds(t, db, bounded); !eqInts(got, 1, 2, 3, 4, 5) {
		t.Errorf("bounded rows = %v, want [1 2 3 4 5]", got)
	}
	if got := mustIds(t, db, unbounded); !eqInts(got, 1, 2, 3, 4, 5) {
		t.Errorf("unbounded rows = %v, want [1 2 3 4 5]", got)
	}

	seek := mustCost(t, db, bounded)
	scan := mustCost(t, db, unbounded)
	// The non-PK correlation re-scans all ~1000 inner rows for each of the 5 outer rows; the PK
	// pushdown seeks instead, so it is an order of magnitude cheaper.
	if seek*10 >= scan {
		t.Errorf("correlated seek %d should be far below the per-outer-row full re-scan %d", seek, scan)
	}
	// Sublinear in the inner size: 5 outer rows, each ≈ a point lookup (height + a row), not ~1000.
	if seek > 400 {
		t.Errorf("correlated seek %d should be sublinear (≈ outer × tree height), not ~5000", seek)
	}
}

func TestCorrelatedScalarSeekMatchesUnboundedRows(t *testing.T) {
	const n = 1000
	db := correlatedTables(t, n)

	// A correlated SCALAR subquery seeking the inner PK returns each outer row's inner value. Rows are
	// identical to what a full re-scan would produce; only the cost differs.
	const bounded = "SELECT (SELECT inr.v FROM inr WHERE inr.id = o.k) FROM o ORDER BY o.id"
	const unbounded = "SELECT (SELECT inr.v FROM inr WHERE inr.v = o.k) FROM o ORDER BY o.id"
	if got := mustIds(t, db, bounded); !eqInts(got, 100, 300, 500, 700, 900) {
		t.Errorf("scalar bounded rows = %v, want [100 300 500 700 900]", got)
	}

	seek := mustCost(t, db, bounded)
	scan := mustCost(t, db, unbounded)
	if seek*10 >= scan {
		t.Errorf("correlated scalar seek %d should be far below the full re-scan %d", seek, scan)
	}
}

func TestCorrelatedMissAndNullOuterSeekNothing(t *testing.T) {
	const n = 1000
	db := correlatedTables(t, n)

	// An outer k with no matching inner id is a point-lookup miss (visits the leaf, reads no row); a
	// NULL outer k is a 3VL-empty bound (reads no page, no row). Neither re-scans the inner.
	if _, err := queryOutcome(db, "INSERT INTO o VALUES (6, 999999), (7, NULL)", nil); err != nil {
		t.Fatal(err)
	}
	const q = "SELECT o.id FROM o WHERE EXISTS (SELECT 1 FROM inr WHERE inr.id = o.k)"
	if got := mustIds(t, db, q); !eqInts(got, 1, 2, 3, 4, 5) {
		t.Errorf("rows = %v, want [1 2 3 4 5] (miss + NULL outer rows drop)", got)
	}
	if seek := mustCost(t, db, q); seek > 500 {
		t.Errorf("seek cost %d should stay sublinear even with a miss and a NULL outer row", seek)
	}
}
